### PR TITLE
alter_table: Add design doc

### PR DESCRIPTION
**Pull request by ParkMyCar**
# [Rendered](https://github.com/ParkMyCar/materialize/blob/alter_table/design-doc/doc/developer/design/20241015_add_columns_to_tables.md)

### Motivation

Adds the design doc that was previously discussed in [Notion](https://www.notion.so/materialize/Adding-Columns-to-Tables-in-Materialize-11213f48d37b806280cbf2d9aa4b5f9a?pvs=4).

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by ParkMyCar** [comment]
@antiguru, @petrosagg, @jkosh44. This is the same design doc from Notion, but with the discussion from the comments added in! Would appreciate any more feedback y'all have, and a stamp when it looks good :)

File: doc/developer/design/20241015_add_columns_to_tables.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+The Storage Controller will need to continue to use `GlobalId`s for operations like rendering a
+source, but it will also need to have some careful management of Persist Handles, e.g. if there are
+two open `WriteHandle`s to the same Persist Shard, writes to one of them would implicitly advance
+the frontier of the other. Or, dropping a `GlobalId` will need to prevent finalizing the underlying
+Persist Shard, if there are other `GlobalId`s that still reference said shard.
+
+### Compute
+
+Our Compute layer will operate entirely on `GlobalId`s. Other than some refactoring of what Catalog
+APIs our Compute layer uses, I don’t anticipate any material changes here.
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply doc/developer/design/20241015_add_columns_to_tables.md]: Good callout, I updated this section to call out we should provide notices to users for this case

File: doc/developer/design/20241015_add_columns_to_tables.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+In planning (or possibly name resolution) is where we will convert from the `CatalogItemId(u1)`
+and `VERSION` syntax in `create_sql` to `GlobalId`s.
+
+In the durable Catalog we will reuse the existing id allocator that currently mints `GlobalId`s to
+mint `CatalogItemId`s. We will create a new id allocator specifically for `GlobalId`s that will be
+initialized to the same value as the original allocator, this prevents accidental `GlobalId` re-use
+if they are persisted outside the Catalog. Additionally we will extend the existing
+[ItemValue](https://github.com/MaterializeInc/materialize/blob/b579caa68b6d287426dead8626c0adc885205740/src/catalog/protos/objects.proto#L119-L126)
+protobuf type to include a map of `VERSION -> GlobalId`. Externally to map between `CatalogItemId`s
+and `GlobalId`s we’ll introduce a new Catalog table, `mz_internal.mz_collection_ids`.
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply doc/developer/design/20241015_add_columns_to_tables.md]: Updated! Let me know if what I wrote makes sense

File: doc/developer/design/20241015_add_columns_to_tables.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+```rust
+enum CatalogItemId {
+  // System namespace.
+  System(u64),
+  // User namespace.
+  User(u64),
+}
+```
+
+### Relationships
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply doc/developer/design/20241015_add_columns_to_tables.md:111]: Totally! I chatted with some Persist folks about this, specifically the use of `CriticalSinceHandle`s in the storage-controller. Once I have something more concrete to test I'll check in with @teskje

**Review by antiguru — APPROVED** [review]
Thanks!

**Comment by ParkMyCar** [comment]
Thanks for the reviews!

**Pull request merged**